doc(kb): workaround to cleanup orphaned instances without orphan cr (#1112)
longhorn/website: content/kb/troubleshooting-orphan-instance.md

---
title: "Troubleshooting: Orphan Engine Or Replica Instance"
authors:
- "Raphanus Lo"
draft: false
date: 2025-05-19
versions:
- "All versions"
categories:
- "orphan"
---

## Applicable versions

* All longhorn versions

## Symptoms

> There are instances, include engine and replica instances, listed in `instancemanager` custom resource, but no associated engine or replica object live in the cluster.

Reviewer: ```suggestion
The `instancemanager` custom resource lists instances (including engine and replica instances), but the corresponding engine or replica objects no longer exist in the cluster.
```

## Details

> In the event of unexpected node disconnections or interruptions during the Longhorn upgrade process, it is possible for the runtime resources of an engine or replica to become decoupled from their corresponding engine or replica custom resources. These resources are called orphaned instances. Orphaned instances can be detected by the instance manager and will be listed in the status of the instance manager. However, they cannot be removed, which may impede normal system upgrades or node replacement processes. This is because the instance manager can only be shut down, restarted, or upgraded once all instances under its management have been safely cleaned up.
> 
> Although Longhorn version 1.9.0 and later support an orphaned instance tracking mechanism, there may still be cases where orphaned resources cannot be tracked or removed if the running instance manager is too outdated. This limitation can further complicate system maintenance and upgrade procedures.

Reviewer: ```suggestion
During unexpected node disconnections or interruptions (such as during a Longhorn upgrade), engine or replica runtime instances can become detached from their corresponding custom resources. These are known as **orphaned instances**.

The Instance Manager can detect orphaned instances and will list them in its status. However, it cannot remove them through standard cleanup procedures. This can block normal operations like system upgrades or node replacements, since the Instance Manager cannot be shut down, restarted, or upgraded until all managed instances are safely cleaned up.

Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these resources. However, if the Instance Manager itself is too outdated, it may not be able to track or remove orphaned instances. This limitation can further complicate maintenance and upgrade processes.
```

## Workaround

> 1. List the engine and replica instances on the node (`worker-node1` in the following example)

Reviewer: ```suggestion
1. List the engine and replica instances on the node (In this example, the node is `worker-node1`.)
```

    Example:
    ```bash
    # kubectl -n longhorn-system get instancemanager -o yaml -l 'longhorn.io/data-engine=v1,longhorn.io/node=worker-node1'

    apiVersion: v1
    items:
    - apiVersion: longhorn.io/v1beta2
      kind: InstanceManager
      metadata:
        labels:
          ...
          longhorn.io/node: worker-node1
        name: instance-manager-8a88a7dd35eab21f30ec566737e87dd0
        namespace: longhorn-system
        ...
      spec:
        nodeID: worker-node1
        ...
      status:
        ...
        instanceEngines:
          orphan-engine-01-e-0:
            spec:
              dataEngine: v1
              name: example-orphan-engine-01-e-0
            ...
        instanceReplicas:
          orphan-replica-01-r-0:
            spec:
              dataEngine: v1
              name: example-orphan-replica-01-r-0
            ...
    ```

> 1. Check the existence of corresponding engine or replica custom resource to identify orphaned instances (engine `example-orphan-engine-01-e-0` and replica `example-orphan-replica-01-r-0` in the following example)

Reviewer: ```suggestion
1. Check the existence of corresponding engine or replica custom resource to identify orphaned instances (In this example, the engine is `example-orphan-engine-01-e-0` and the replica is`example-orphan-replica-01-r-0`).
```

    Example:
    ```bash
    # kubectl -n longhorn-system get engine example-orphan-engine-01-e-0
    error: the server doesn't have a resource type "engine"

    # kubectl -n longhorn-system get replica example-orphan-replica-01-r-0
    error: the server doesn't have a resource type "replica"
    ```

> 1. Shell into the pod of instance manager. The name of the pod is identical with the instance manager (`instance-manager-8a88a7dd35eab21f30ec566737e87dd0` in the following example)

Reviewer: ```suggestion
1. Shell into the pod whose name matches the Instance Manager (In this example, the instance manager is `instance-manager-8a88a7dd35eab21f30ec566737e87dd0`).
```

    Example:
    ```
    # kubectl -n longhorn-system exec -i -t instance-manager-8a88a7dd35eab21f30ec566737e87dd0 -- /bin/bash
    ```

> 1. Inside the instance manager pod, delete the instances using `instance-manager` CLI tool.

Reviewer: ```suggestion
1. Use the `instance-manager` CLI tool to remove the orphaned engine and replica instances.
```

    ```bash
    # instance-manager process delete --name example-orphan-engine-01-e-0
    # instance-manager process delete --name example-orphan-replica-01-r-0
    ```

> 1. Confirm the instance is cleared from the instance manager.

Reviewer: ```suggestion
1. Confirm that the instances are no longer listed in the Instance Manager status.
```

---
title: "Troubleshooting: Orphan Engine Or Replica Instance"
authors:
- "Raphanus Lo"
draft: false
date: 2025-05-19
versions:
- "All versions"
categories:
- "orphan"
---

## Applicable versions

* All longhorn versions

## Symptoms

> The `instancemanager` custom resource lists instances (including engine and replica instances), but the corresponding engine or replica objects no longer exist in the cluster.

Reviewer 2: ```suggestion
The `instancemanager` custom resource (CR) lists runtime instances (including engine and replica runtime instances), but the corresponding engine or replica CRs no longer exist in the cluster.
```

## Details

> During unexpected node disconnections or interruptions (such as during a Longhorn upgrade), engine or replica runtime instances can become detached from their corresponding custom resources. These are known as **orphaned instances**.

Reviewer 2: ```suggestion
During unexpected node disconnections or interruptions (such as during a Longhorn upgrade), engine or replica runtime instances can become detached from their corresponding CRs. These are known as **orphaned instances**.
```

The Instance Manager can detect orphaned instances and will list them in its status. However, it cannot remove them through standard cleanup procedures. This can block normal operations like system upgrades or node replacements, since the Instance Manager cannot be shut down, restarted, or upgraded until all managed instances are safely cleaned up.

> Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these orphans. However, if the Instance Manager is created by Longhorn before v1.9.0, it is not able to track or remove orphaned instances.

Reviewer: ```suggestion
Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these orphaned instances. However, if the Instance Manager is created by Longhorn before v1.9.0, it is not able to track or remove orphaned instances.
```

> Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these orphans. However, if the Instance Manager is created by Longhorn before v1.9.0, it is not able to track or remove orphaned instances.

Reviewer 2: ```suggestion
Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these orphans. However, if an Instance Manager was created by a Longhorn version earlier than v1.9.0, it will not be able to track or remove orphaned instances. In such cases, you can manually remove them by applying the provided workaround.
```

Reviewer: Thanks for the suggestion. One more thing, we should write `these orphaned instances` in places of `orphans`

Author: Reworded per discussion with @sushant-suse :
> Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these orphaned instances. However, if an Instance Manager was created by a Longhorn version earlier than v1.9.0, it will not be able to track or remove orphaned instances. In such cases, you can manually remove them by following the provided workaround.

Reviewer: LGTM.

> Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these resources. However, if the Instance Manager itself is too outdated, it may not be able to track or remove orphaned instances. This limitation can further complicate maintenance and upgrade processes.

Reviewer 2: Not really sure what "This limitation can further complicate maintenance and upgrade processes."  means here.

```suggestion
Starting with Longhorn v1.9.0, an orphaned instance tracking mechanism is introduced to help manage these orphans. However, if the Instance Manager is created by Longhorn before v1.9.0, it is not able to track or remove orphaned instances. 
```

Reviewer 2: Need to mention what the purpose of the following workaround is.

## Workaround

1. List the engine and replica instances on the node (In this example, the node is `worker-node1`.)
    Example:
    ```bash
    # kubectl -n longhorn-system get instancemanager -o yaml -l 'longhorn.io/data-engine=v1,longhorn.io/node=worker-node1'

    apiVersion: v1
    items:
    - apiVersion: longhorn.io/v1beta2
      kind: InstanceManager
      metadata:
        labels:
          ...
          longhorn.io/node: worker-node1
        name: instance-manager-8a88a7dd35eab21f30ec566737e87dd0
        namespace: longhorn-system
        ...
      spec:
        nodeID: worker-node1
        ...
      status:
        ...
        instanceEngines:
          orphan-engine-01-e-0:
            spec:
              dataEngine: v1
              name: example-orphan-engine-01-e-0
            ...
        instanceReplicas:
          orphan-replica-01-r-0:
            spec:
              dataEngine: v1
              name: example-orphan-replica-01-r-0
            ...
    ```

> 1. Check the existence of corresponding engine or replica custom resource to identify orphaned instances (In this example, the engine is `example-orphan-engine-01-e-0` and the replica is`example-orphan-replica-01-r-0`).

Reviewer 2: ```suggestion
1. Check the existence of the corresponding engine or replica CR to identify orphaned instances (In this example, the engine is `example-orphan-engine-01-e-0` and the replica is`example-orphan-replica-01-r-0`).
```

    Example:
    ```
    # kubectl -n longhorn-system exec -i -t instance-manager-8a88a7dd35eab21f30ec566737e87dd0 -- /bin/bash
    ```
1. Use the `instance-manager` CLI tool to remove the orphaned engine and replica instances.
    ```bash
    # instance-manager process delete --name example-orphan-engine-01-e-0
    # instance-manager process delete --name example-orphan-replica-01-r-0
    ```
1. Confirm that the instances are no longer listed in the Instance Manager status.
    ```bash
    # kubectl -n longhorn-system get instancemanager -o yaml -l 'longhorn.io/data-engine=v1,longhorn.io/node=worker-node1'

    apiVersion: v1
    items:
    - apiVersion: longhorn.io/v1beta2
      kind: InstanceManager
      metadata:
        labels:
          ...
          longhorn.io/node: worker-node1
        name: instance-manager-8a88a7dd35eab21f30ec566737e87dd0
        namespace: longhorn-system
        ...
      spec:
        nodeID: worker-node1
        ...
      status:
        ...(engine and replica are removed)..
    ```

## Related information

> - Related Longhorn issue: https://github.com/longhorn/longhorn/issues/6764

Reviewer: ```suggestion
- Related Longhorn issue [#6764](https://github.com/longhorn/longhorn/issues/6764).
```
